File: src/grove/manager.py
import os
import shutil
from pathlib import Path
from typing import List, Optional
from grove.config import ConfigManager
from grove import git

class GroveManager:
    def __init__(self):
        self.config = ConfigManager()
        self.cache_dir = Path.home() / ".grove" / "cache"

    # --- Set Operations ---
    def add_set(self, name: str, repos: List[str]):
        """Define a new set of repositories."""
        if self.config.get_set(name):
            raise ValueError(f"Set '{name}' already exists.")
        
        self.config.add_set(name, repos)
        print(f"Set '{name}' created with {len(repos)} repositories.")

    def update_set(self, name: str, new_name: Optional[str] = None, 
                   add_repos: Optional[List[str]] = None, 
                   remove_repos: Optional[List[str]] = None):
        self.config.update_set(name, new_name, add_repos, remove_repos)
        print(f"Set updated.")

    def remove_set(self, name: str):
        self.config.remove_set(name)
        print(f"Set '{name}' removed.")

    # --- Feature Operations ---
    def create_feature(self, set_name: str, feature_name: str):
        """Create a new worktree feature for a given set."""
        set_data = self.config.get_set(set_name)
        if not set_data:
            raise ValueError(f"Set '{set_name}' not found.")
        
        if self.config.get_feature(feature_name):
            raise ValueError(f"Feature '{feature_name}' already exists.")
        
        root_dir = self.config.get_root_dir()
        feature_path = root_dir / set_name / feature_name
        
        if feature_path.exists():
             raise ValueError(f"Directory {feature_path} already exists. Please manually remove it or choose a different name.")

        print(f"Creating feature '{feature_name}' for set '{set_name}' at {feature_path}...")
        feature_path.mkdir(parents=True)

        for repo_url in set_data["repos"]:
            try:
                # 1. Ensure bare repo
                bare_repo = git.ensure_bare_repo(repo_url, self.cache_dir)
                
                # 2. Determine inner path (repo name)
                repo_name = git.get_repo_name_from_url(repo_url)
                target_path = feature_path / repo_name
                
                # 3. Create worktree
                print(f"  Adding worktree for {repo_name}...")
                git.create_worktree(bare_repo, feature_name, target_path)
                
            except Exception as e:
                print(f"  Error processing {repo_url}: {e}")
                # TODO: Cleanup logic if partial failure?
                
        # 4. Initialize Skills
        self._init_skills(set_data.get("skills_dir"), feature_path)

        # 5. Register in config
        self.config.add_feature(feature_name, set_name, str(feature_path))
        print(f"Feature '{feature_name}' ready.")

    def remove_feature(self, feature_name: str):
        feature_data = self.config.get_feature(feature_name)
        if not feature_data:
            raise ValueError(f"Feature '{feature_name}' not found.")
        
        path = Path(feature_data["path"])
        set_name = feature_data["set"]
        set_data = self.config.get_set(set_name)
        
        print(f"Removing feature '{feature_name}' at {path}...")
        
        # 1. Remove worktrees
        # We need to cleanup the bare repos basically. 
        # The cleanest is to go into each repo dir and `git worktree remove .`? 
        # But we deleted the lines in git.py.
        # Let's retry: for each repo in the Set, go to its bare repo and remove the worktree matching this path.
        if set_data: 
             for repo_url in set_data["repos"]:
                repo_name = git.get_repo_name_from_url(repo_url)
                bare_repo = self.cache_dir / repo_name
                target_path = path / repo_name
                
                if bare_repo.exists():
                     try:
                         # Force remove the worktree reference
                         git.run_git(["worktree", "remove", "--force", str(target_path)], cwd=bare_repo)
                     except Exception as e:
                         print(f"  Warning: Failed to remove worktree for {repo_name}: {e}")

        # 2. Delete directory (if allowed)
        if path.exists():
            shutil.rmtree(path)
            
        # 3. Update config
        self.config.remove_feature(feature_name)
        print(f"Feature '{feature_name}' removed.")

    def _init_skills(self, skills_src: Optional[str], feature_path: Path):
        """Initialize skills in the feature directory."""
        if not skills_src:
            return

        src_path = Path(skills_src).expanduser()
        if not src_path.exists():
            # If source doesn't exist, maybe we just create the dest folder empty?
            # Or just warn.
            return 

        dest_path = feature_path / ".agent" / "skills"
        dest_path.parent.mkdir(parents=True, exist_ok=True)
        
        if dest_path.exists():
            shutil.rmtree(dest_path)
            
        # Join the skills from the set to the local skills
        # We can symlink or copy. Copy is safer for now to avoid side effects? 
        # User said "grove_dir/worktreesetB/.gemini/skills".
        # Let's copy.
        shutil.copytree(src_path, dest_path)
        print(f"  Skills initialized from {src_path}")

File: internal/manager/manager.go
package manager

import (
	"fmt"
	"github.com/vedantprajapati/Grove/internal/config"
	"github.com/vedantprajapati/Grove/internal/git"
	"io"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

type Manager struct {
	Config   *config.Config
	CacheDir string // Directory for bare repos
}

func NewManager() (*Manager, error) {
	cfg, err := config.LoadConfig("")
	if err != nil {
		return nil, err
	}
	home, _ := os.UserHomeDir()
	return &Manager{
		Config:   cfg,
		CacheDir: filepath.Join(home, ".grove", "cache"),
	}, nil
}

func NewManagerWithConfig(path string) (*Manager, error) {
	cfg, err := config.LoadConfig(path)
	if err != nil {
		return nil, err
	}
	home, _ := os.UserHomeDir()
	return &Manager{
		Config:   cfg,
		CacheDir: filepath.Join(home, ".grove", "cache"),
	}, nil
}

func (m *Manager) SaveConfig() error {
	return m.Config.Save()
}

// --- Set Operations ---

func (m *Manager) AddSet(name string, repos []string) error {
	if _, exists := m.Config.Sets[name]; exists {
		return fmt.Errorf("set '%s' already exists", name)
	}

	home, _ := os.UserHomeDir()
	skillsDir := filepath.Join(home, ".grove", "skills", name)

	m.Config.Sets[name] = config.Set{
		Repos:     repos,
		SkillsDir: skillsDir,
	}
	return m.SaveConfig()
}

func (m *Manager) RemoveSet(name string) error {
	if _, exists := m.Config.Sets[name]; !exists {
		return fmt.Errorf("set '%s' not found", name)
	}

	// Check if any feature uses this set
	for _, feat := range m.Config.Features {
		if feat.Set == name {
			return fmt.Errorf("cannot remove set '%s': used by active feature", name)
		}
	}

	delete(m.Config.Sets, name)
	return m.SaveConfig()
}

// --- Feature Operations ---

func (m *Manager) CreateFeature(setName, featureName string) error {
	set, ok := m.Config.Sets[setName]
	if !ok {
		return fmt.Errorf("set '%s' not found", setName)
	}

	if _, exists := m.Config.Features[featureName]; exists {
		return fmt.Errorf("feature '%s' already exists", featureName)
	}

	// Prepare directories
	rootDir, err := config.ExpandPath(m.Config.RootDir)
	if err != nil {
		return err
	}

	// Structure: root_dir/set/feature
	featurePath := filepath.Join(rootDir, setName, featureName)
	if _, err := os.Stat(featurePath); !os.IsNotExist(err) {
		return fmt.Errorf("directory %s already exists", featurePath)
	}

	cacheDir := m.CacheDir

	fmt.Printf("Creating feature '%s' for set '%s' at %s...\n", featureName, setName, featurePath)

	if err := os.MkdirAll(featurePath, 0755); err != nil {
		return fmt.Errorf("failed to create feature directory: %v", err)
	}

	// 1. Git Operations (Parallel)
	var wg sync.WaitGroup
	errChan := make(chan error, len(set.Repos))

	for _, repoURL := range set.Repos {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			bareRepo, err := git.EnsureBareRepo(url, cacheDir)
			if err != nil {
				errChan <- fmt.Errorf("failed to ensure bare repo for %s: %v", url, err)
				return
			}

			repoName := git.GetRepoNameFromURL(url)
			targetPath := filepath.Join(featurePath, repoName)

			fmt.Printf("Adding worktree for %s...\n", repoName)
			if err := git.CreateWorktree(bareRepo, featureName, targetPath); err != nil {
				errChan <- err
				return
			}
		}(repoURL)
	}

	wg.Wait()
	close(errChan)

	// Collect first error if any
	for err := range errChan {
		if err != nil {
			return err
		}
	}

	// 2. Skills Initialization
	if err := m.initSkills(set, rootDir, setName); err != nil {
		fmt.Printf("Warning: Failed to init skills: %v\n", err)
	}

	// 3. Update Config
	m.Config.Features[featureName] = config.Feature{
		Path: featurePath,
		Set:  setName,
	}

	return m.SaveConfig()
}

func (m *Manager) SyncFeature(featureName string) error {
	feat, ok := m.Config.Features[featureName]
	if !ok {
		return fmt.Errorf("feature '%s' not found", featureName)
	}

	set, ok := m.Config.Sets[feat.Set]
	if !ok {
		return fmt.Errorf("set '%s' not found for feature", feat.Set)
	}

	fmt.Printf("Syncing feature '%s' (Set: %s) in parallel...\n", featureName, feat.Set)

	var wg sync.WaitGroup
	errChan := make(chan error, len(set.Repos))

	for _, repoURL := range set.Repos {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			repoName := git.GetRepoNameFromURL(url)
			repoPath := filepath.Join(feat.Path, repoName)
			if err := git.SyncRepo(repoPath); err != nil {
				errChan <- fmt.Errorf("error syncing %s: %v", repoName, err)
			} else {
				fmt.Printf("  %s synced.\n", repoName)
			}
		}(repoURL)
	}

	wg.Wait()
	close(errChan)

	var errors []string
	for err := range errChan {
		errors = append(errors, err.Error())
	}

	if len(errors) > 0 {
		return fmt.Errorf("sync failed for some repositories:\n%s", strings.Join(errors, "\n"))
	}

	return nil
}

func (m *Manager) ExecFeature(featureName string, command string, args []string) error {
	feat, ok := m.Config.Features[featureName]
	if !ok {
		return fmt.Errorf("feature '%s' not found", featureName)
	}

	set, ok := m.Config.Sets[feat.Set]
	if !ok {
		return fmt.Errorf("set '%s' not found for feature", feat.Set)
	}

	fmt.Printf("Executing '%s %s' across %d repos...\n", command, strings.Join(args, " "), len(set.Repos))

	var wg sync.WaitGroup
	for _, repoURL := range set.Repos {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			repoName := git.GetRepoNameFromURL(url)
			repoPath := filepath.Join(feat.Path, repoName)

			fmt.Printf("\n--- [%s] ---\n", repoName)
			output, err := git.RunCommand(repoPath, command, args...)
			if err != nil {
				fmt.Printf("Error in %s: %v\nOutput: %s\n", repoName, err, output)
			} else {
				fmt.Println(output)
			}
		}(repoURL)
	}

	wg.Wait()
	return nil
}

type RepoStatus struct {
	Name    string
	Branch  string
	IsDirty bool
	ABC     string // Ahead/Behind Count
}

func (m *Manager) GetFeatureStatus(featureName string) ([]RepoStatus, error) {
	feat, ok := m.Config.Features[featureName]
	if !ok {
		return nil, fmt.Errorf("feature '%s' not found", featureName)
	}

	set, ok := m.Config.Sets[feat.Set]
	if !ok {
		return nil, fmt.Errorf("set '%s' not found for feature", feat.Set)
	}

	var wg sync.WaitGroup
	statusChan := make(chan RepoStatus, len(set.Repos))

	for _, repoURL := range set.Repos {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			repoName := git.GetRepoNameFromURL(url)
			repoPath := filepath.Join(feat.Path, repoName)

			branch, _ := git.BranchName(repoPath)
			isDirty, abc, _ := git.GetStatus(repoPath)

			statusChan <- RepoStatus{
				Name:    repoName,
				Branch:  branch,
				IsDirty: isDirty,
				ABC:     abc,
			}
		}(repoURL)
	}

	wg.Wait()
	close(statusChan)

	var statuses []RepoStatus
	for s := range statusChan {
		statuses = append(statuses, s)
	}
	return statuses, nil
}

func (m *Manager) RemoveFeature(featureName string) error {
	feat, ok := m.Config.Features[featureName]
	if !ok {
		return fmt.Errorf("feature '%s' not found", featureName)
	}

	fmt.Printf("Removing feature '%s'...\n", featureName)

	// 1. Remove Worktrees (Parallel cleanup)
	cacheDir := m.CacheDir
	if set, ok := m.Config.Sets[feat.Set]; ok {
		var wg sync.WaitGroup
		for _, repoURL := range set.Repos {
			wg.Add(1)
			go func(url string) {
				defer wg.Done()
				repoName := git.GetRepoNameFromURL(url)
				bareRepo := filepath.Join(cacheDir, repoName)
				worktreeStr := filepath.Join(feat.Path, repoName)
				if err := git.RemoveWorktree(bareRepo, worktreeStr); err != nil {
					fmt.Printf("  Warning: failed to clean worktree for %s: %v\n", repoName, err)
				}
			}(repoURL)
		}
		wg.Wait()
	}

	// 2. Remove Directory
	if err := os.RemoveAll(feat.Path); err != nil {
		return fmt.Errorf("failed to remove directory %s: %v", feat.Path, err)
	}

	// 3. Update Config
	delete(m.Config.Features, featureName)
	return m.SaveConfig()
}

func (m *Manager) initSkills(set config.Set, rootDir, setName string) error {
	// Destination: root_dir/set/.gemini/skills (Shared for the set)
	// Actually, user said: grove_dir/worktreesetA/.gemini/skills
	// And features are siblings?
	// User said:
	// grove_dir/worktreesetA/feature1
	// grove_dir/worktreesetA/feature2
	// grove_dir/worktreesetA/.gemini/skills

	destDir := filepath.Join(rootDir, setName, ".gemini", "skills")

	// If source skills dir is defined and exists, copy it
	srcDir, err := config.ExpandPath(set.SkillsDir)
	if err != nil {
		return err
	}

	if _, err := os.Stat(srcDir); os.IsNotExist(err) {
		// Source doesn't exist, skip
		return nil
	}

	// Remove old if exists (re-sync)
	os.RemoveAll(destDir)

	return copyDir(srcDir, destDir)
}

// copyDir recursively copies a directory tree, attempting to preserve permissions.
// Source directory must exist.
func copyDir(src string, dst string) error {
	src = filepath.Clean(src)
	dst = filepath.Clean(dst)

	si, err := os.Stat(src)
	if err != nil {
		return err
	}
	if !si.IsDir() {
		return fmt.Errorf("source is not a directory")
	}

	err = os.MkdirAll(dst, si.Mode())
	if err != nil {
		return err
	}

	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())

		if entry.IsDir() {
			err = copyDir(srcPath, dstPath)
			if err != nil {
				return err
			}
		} else {
			// Copy file
			err = copyFile(srcPath, dstPath)
			if err != nil {
				return err
			}
		}
	}

	return nil
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	if err != nil {
		return err
	}
	return out.Close()
}
